File: stepik-union/src/main/java/org/stepik/core/courseFormat/StepType.java
package org.stepik.core.courseFormat;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * @author meanmail
 */
public enum StepType {
    UNKNOWN("unknown"),
    CODE("code"),
    TEXT("text"),
    VIDEO("video"),
    CHOICE("choice"),
    STRING("string"),
    SORTING("sorting"),
    MATCHING("matching"),
    NUMBER("number"),
    DATASET("dataset"),
    TABLE("table"),
    FILL_BLANKS("fill-blanks"),
    MATH("math"),
    FREE_ANSWER("free-answer");

    private static Map<String, StepType> map;
    private final String name;

    StepType(String name) {
        this.name = name;
    }

    @NotNull
    public static StepType of(String name) {
        if (map == null) {
            map = new HashMap<>();
            Arrays.stream(values())
                    .forEach(value -> map.put(value.name, value));
        }
        return map.getOrDefault(name, UNKNOWN);
    }

    @Override
    public String toString() {
        return name;
    }

    public String getName() {
        return name;
    }
}

File: stepik-union/src/main/kotlin/org/stepik/core/courseFormat/StepType.kt
package org.stepik.core.courseFormat


enum class StepType(val typeName: String) {
    UNKNOWN("unknown"),
    CODE("code"),
    TEXT("text"),
    VIDEO("video"),
    CHOICE("choice"),
    STRING("string"),
    SORTING("sorting"),
    MATCHING("matching"),
    NUMBER("number"),
    DATASET("dataset"),
    TABLE("table"),
    FILL_BLANKS("fill-blanks"),
    MATH("math"),
    FREE_ANSWER("free-answer");

    override fun toString(): String {
        return typeName
    }

    companion object {

        private val map: Map<String, StepType> by lazy {
            values().map { it.typeName to it }.toMap()
        }

        fun of(typeName: String): StepType {
            return map[typeName] ?: UNKNOWN
        }
    }
}
